### PR TITLE
Update standards based on Nov 2019 PyHC meeting discussion

### PR DESCRIPTION
**Pull request by namurphy**
This pull request is to include changes to the community standards based on the discussion at the PyHC meeting from November 2019.  We probably still need to update the author list to reflect the people who were in attendance in Boulder.  We'll need to review this during a PyHC telecon before we accept this, which will probably be in January 2020.

Here are the notes from the meeting back in November:

 - Change: All functions, classes, and modules **in the public-facing API**...
 - Question: what do people think of [Black](https://pypi.org/project/black/)?
    - It’s opinionated and consistent with the standards
 - Add actual links for “System[link]” and “acceptance[link]”
 - Python 2.7 has about a thousand hours left
 - Change: “clearly **and constructively** explain when a contribution is not accepted”
 - Change in #15: “Projects should have a reporting procedure for violations of the code of conduct (see [link to NumFOCUS’s for ).”
 - Add a bullet point
    - Be explicit on what versions of dependencies a package supports.
    - Strong recommendation that we should support [NumPy Enhancement Proposal 29](https://numpy.org/neps/nep-0029-deprecation_policy.html) which provides recommendations on which prior versions of Python and NumPy to support.
    - “Should” is fine 
 - Do we want to figure out a way to “grade” packages on the PyHC website?
    - Making an assessment can be really useful to packages
    - Probably need someone(s) to go through packages
    - Goes back to governance
    - Self-assessments are a first step; and are important
        - Maybe have a template for that particular type of PR 
    - Maybe if someone wants to add their project to PyHC’s website, then they should provide the results of their self-assessment
    - [Astropy's affiliated packages list](https://www.astropy.org/affiliated/) has a “traffic light system” (red/orange/green)
    - Would be wonderful to have community reviews; really helpful for packages
        - JOSS is a good model for this
    - Should current packages plan on going through this procedure, doing a self-assessment?
    - We’ll need to first figure out what the items to be evaluated on: what the metrics are
    - Does anyone want to volunteer?
        - Eerie silence
    - Bring up at next telecon
    - **By the next meeting in the spring, aim to have all packages on the website reviewed**

### CODE REVIEW

**Comment by namurphy** [comment]
If you approve these changes and would like to be listed as a co-author, please let me know!  You can add a comment, suggest a commit, or approve this pull request.  It'd be helpful to know the name/affiliation you would like used if they're different from the original document, as well as your ORCID if you have one.

Many thanks to @ehsteve and everyone else who shepherded the process the first time around!

**Comment by namurphy** [comment]
We also have been talking about who is willing to volunteer for going through PyHC packages and evaluating the extent to which each package meets the standards.  We could potentially do this during the hackathon following our spring meeting in Cambridge.  The main meeting will be from April 27-29, with the hackathon beginning on the afternoon of the 29th and lasting ~1-2 days.

**Comment by wtbarnes** [comment]
My affiliation should be updated to U.S. Naval Research Laboratory and my ORCID is 0000-0001-9642-6089.

**Comment by wtbarnes** [comment]
> We also have been talking about who is willing to volunteer for going through PyHC packages and evaluating the extent to which each package meets the standards. We could potentially do this during the hackathon following our spring meeting in Cambridge.

I wonder if this could be a responsibility of the proposed PyHC developer position?

File: standards.md [hunk 1]
```diff
@@ -29,12 +35,12 @@ Definitions:
...
-9. **Testing**: Stable packages must provide unit tests of individual components (e.g. functions, classes) as well as integration tests that test the interaction between components that covers most of the code. Testing coverage should be measured. Automated testing is recommended, in which tests are run before any code is merged. System[link] and acceptance[link] testing are also recommended.
+8. **Documentation**: All functions, classes, and modules in the public-facing application programming interface (API) must have documentation strings (docstrings) provided in a standard [convention](https://www.python.org/dev/peps/pep-0257/) (e.g. [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html)). Docstrings must describe the code’s purpose, describe all inputs and outputs, and provide examples. High level documentation must also be provided as guides, tutorials, and developer docs. Documentation must be provided in version control with the code and be made available online in a readable form.
```
Comment (ehsteve) [inline standards.md]: I'm not sure I understand this change. This suggests that non public-facing functions do not need doc strings. How is a contributor supposed to interact with developer code that is not documented?

File: standards.md [hunk 2]
```diff
@@ -29,12 +35,12 @@ Definitions:
...
+9. **Testing**: Stable packages must provide unit tests of individual components (e.g. functions, classes) as well as integration tests that test the interaction between components that covers most of the code. Testing coverage should be measured. Automated testing is recommended, in which tests are run before any code is merged. System and acceptance testing are also recommended.
 10. **Dependencies**: Projects should import the minimum number of packages necessary. 	Adding new dependencies should be a __considered__ decision.
 11. **Python 3**: All packages must be compatible or work towards being compatible with Python 3. Providing ongoing support for Python 2 is not recommended as the end of life for Python 2 is January 1, 2020 (see [PEP 373](https://www.python.org/dev/peps/pep-0373/)).
```
Comment (ehsteve) [inline standards.md]: I would update this to state that Python 2 has reached its end of life and would rephrase to be more forceful in that all packages MUST support Python 3.

File: standards.md (hunk 1 above)
Comment (namurphy) [reply standards.md]: Hm...good point.  I've occasionally written short private helper functions/methods with no arguments that have their one-line docstring encoded in the function name, so the docstring doesn't end up adding any additional information.  I'm also sometimes lax on docstrings for unstable temporary implementations that I know will not be lasting long.  Perhaps we could say that all public facing functions/etc *must* have docstrings, and that private functions/etc *should* generally have docstrings?

File: standards.md (hunk 2 above)
Comment (namurphy) [reply standards.md]: Agreed completely!  This slipped my mind because I was making these edits in 2019.

**Comment by namurphy** [comment]
> I think it would always be better to have discussions of PRs on GitHub so that they are "in the record" and everyone is able to participate.

Good point!  I had thought I had emailed the list after I originally made the PR, but I forgot to...  I put my notes from today's discussion in the commit log, though I should have put it as a comment here.  For discussions at meetings/telecons, we should make sure that notes are posted in the corresponding PR. 

> One topic of discussion during this telecon was on adoption of NEP 29 (which created a community-wide standard for how long to support older versions of Python and SciPy) by PyHC packages.  Stuart Mumford suggested that it be okay for some packages to have stricter standards, and that packages can support older versions of NumPy and Python if they so choose.  Michael Hirsch brought up that
usage/download statistics from a package that supports Python 3.5+ showed that very few people are using Python 3.5 at the moment (and also that Python 3.5 gives the most trouble for the testing matrix).
I brought up that core packages like SunPy should follow the guidelines of NEP 29 most closely, while packages that are currently under development and have few users would be okay in supporting Python 3.7+ or even 3.8+ if the development phase is going to last ~1-2 years.  We decided that using "should" leaves open enough leeway for us to accept packages that have different requirements.  I forgot
to mention that I think that we should follow the same guidelines for Astropy (and ideally packages within PyHC) as we do for NumPy.
> 
> (These notes are a convolution of what people actually said, and my ability to remember 42 minutes into the past.)

File: standards.md (hunk 2 above)
Comment (scivision) [reply standards.md]: 11. python3 may be redundant with 12. deprecation. It may now (2020) finally be sufficient to just omit mention of python2 and just mention NEP29.

File: standards.md (hunk 1 above)
Comment (aburrell) [reply standards.md]: I think "all public facing functions/etc must have docstrings, and that private functions/etc should generally have docstrings" is a good compromise.  I know there are non-public facing functions in SciPy that use internal comments instead of docstrings when the functions are never supposed to be interacted with externally.

File: standards.md [hunk 3]
```diff
@@ -4,14 +4,20 @@
...
 
-Agreed on 10-Dec-2018 by (in alpahbetical order)
+Agreed on 10-Dec-2018 by (in alphabetical order)
 
 **A. Annex** (JHU), **B. L. Alterman** (Univ. of Michigan), **A. Azari** (Univ. of Michigan), **W. Barnes** (Rice Univ.), **M. Bobra** (Stanford), **B. Cecconi** (Observartoire de Paris), **S. Christe** (NASA GSFC), **J. Coxon** (Univ. of Southampton), **A. DeWolfe** (LASP), **A. Halford** (Aerospace Corporation), **B. Harter** (LASP), **J. Ireland** (NASA GSFC), **J. Jahn** (SwRI), **J. Klenzing** (NASA GSFC), **M. Liu** (SunPy), **J. Mason** (NASA GSFC), **R. McGranaghan** (NASA JPL), **N. Murphy** (CfA), **S. Murray** (Trinity College Dublin), **J. Niehof** (Univ. of New Hampshire), **M.D. Nguyen** (Lomonosov Moscow State Univ.), **R. Panneton** (CU/LASP), **A. Pembroke** (NASA GSFC), **D. Pérez-Suárez** (University College London), **C. Piker** (Univ. of Iowa), **A. Roberts** (NASA GSFC), **D. Ryan** (NASA GSFC), **S. Savage** (NASA GSFC), **J. Smith** (NASA GSFC, Catholic Univ.), **D. Stansby** (Imperial College London), **J. Vandegriff** (JHU/APL), **R. S. Weigel** (George Mason University)
 
+Amended on 06-Jan-2020 by (in alphabetical order)
+
+**N. Murphy** (CfA), 
+
```
Comment (aburrell) [inline standards.md]: **A. G. Burrell** (NRL).  Also, A. Halford is now at NASA GSFC.

**Comment by scivision** [comment]
### minimum Python version observations

I have required Python 3.6 for over a year on almost 100% of my packages with only a handful of complaints, including from incorporation into Linux distributions, offline/proprietary/OEM users and others not captured by PyPi. 

* PyMap3D (230K total PyPi downloads) has < 1% Python 3.5 PyPi downloads. PyMap3D still supports Python 3.5 to support OEM with long dev cycle on embedded systems.
* Numpy is down to about 15% Python 3.5 PyPi downloads, and that may start dropping soon as Python 3.5 won't be supported in the next Numpy release.

Although there are limitations on relevance of PyPi download stats (does not account for `conda` and `apt install` etc), this script allows easy plotting of last 6 months of PyPi download data over time: https://github.com/scivision/pypistats-plots

As Jeff also noted, there are increasing penalties to supporting Python 3.5. If one looks at and assumes sufficient representation of PyPi stats, Python 3.5 may be a small audience for a growing number of packages.

File: standards.md [hunk 4]
```diff
@@ -29,12 +35,12 @@ Definitions:
...
-15. **Code of conduct**:  Each project must adopt a code of conduct that is compatible with the [Contributor Covenant](https://www.contributor-covenant.org) and make it publicly available.
-
+12. **Deprecation Policy** (in accordance with [NEP 29](https://numpy.org/neps/nep-0029-deprecation_policy.html)): Each project should support (1) all minor versions of Python released 42 months prior to the project, and at minimum the two latest minor versions; and (2) all minor versions of NumPy released in the 24 months prior to the project, and at minimum the last three minor versions.  In ``setup.py``, the ``python_requires`` variable should be set to the minimum supported version of Python.  All supported minor versions of Python should be in the test matrix and have binary artifacts built for the release.  Minimum Python and NumPy version support should be adjusted upward only on major and minor releases, and never on a patch release.
```
Comment (scivision) [inline standards.md]: ...In `setup.py` or `setup.cfg`, the `python_requires` ...

Although perhaps too technical for the standards document, in my opinion setup.cfg should be used as much as possible to allow better tool introspection without needing to install. For security, best practices, etc.

**Comment by scivision** [comment]
Re: Black, I think it may be too much to have in PyHC standards in general, I have run into its corner cases that make code less readable. Also the whole defaulting to double quotes thing is problematic to me, although `black -S` is a workaround to retain single quotes.

I would rather have a statement encouraging setup.cfg over setup.py if we're going to get technical ;-)

**Comment by scivision** [comment]
re: Docs
 in my opinion this should include as OK even those that use non-readthedocs sites like GitHub Pages. I don't recall the source, but I read several months ago that packages of mine that do use `numpydoc` extensively with GitHub Pages were marked as **not** having docs because I don't use readthedocs.

**Comment by scivision** [comment]
my ORCID: 0000-0002-1637-6526

File: standards.md (hunk 4 above)
Comment (dpshelio) [reply standards.md]: Agree with @scivision. We should encourage to keep that metadata on `setup.cfg` or `pyproject.toml`.

File: standards.md (hunk 1 above)
Comment (Cadair) [reply standards.md]: Certainly requiring full numpydoc docstrings on all private functions is overkill. +1 on @aburrell 's suggestion

Comment (ehsteve) [reply standards.md]: I'm fine with going with the "should". I don't really like the idea of encouraging "second-class" functions that are more poorly documented than others but I understand the concern here. I do know that for sunpy there are *many* important non-public facing functions that need good documentation for developers to interact with them!

**Comment by blalterman** [comment]
Please add me to the list. Orcid is 0000-0001-6673-3432

File: standards.md [hunk 5]
```diff
@@ -29,12 +35,12 @@ Definitions:
...
-9. **Testing**: Stable packages must provide unit tests of individual components (e.g. functions, classes) as well as integration tests that test the interaction between components that covers most of the code. Testing coverage should be measured. Automated testing is recommended, in which tests are run before any code is merged. System[link] and acceptance[link] testing are also recommended.
+8. **Documentation**: All functions, classes, and modules in the public-facing application programming interface (API) must have documentation strings (docstrings) provided in a standard [convention](https://www.python.org/dev/peps/pep-0257/) (e.g. [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html)).  Functions, classes, and modules that are not public-facing should generally have docstring to enable developers to maintain them.  Docstrings must describe the code’s purpose, describe all inputs and outputs, and provide examples. High level documentation must also be provided as guides, tutorials, and developer docs. Documentation must be provided in version control with the code and be made available online in a readable form.
```
Comment (jtniehof) [inline standards.md]: "should generally have docstring" --should be plural.

File: standards.md [hunk 6]
```diff
@@ -29,12 +35,12 @@ Definitions:
...
-15. **Code of conduct**:  Each project must adopt a code of conduct that is compatible with the [Contributor Covenant](https://www.contributor-covenant.org) and make it publicly available.
-
+11. **Python 3**: All packages must be compatible with Python 3. Providing ongoing support for Python 2 is not recommended as the end of life for Python 2 was January 1, 2020 (see [PEP 373](https://www.python.org/dev/peps/pep-0373/)).
+12. **Deprecation Policy** (in accordance with [NEP 29](https://numpy.org/neps/nep-0029-deprecation_policy.html)): Each project should support (1) all minor versions of Python released 42 months prior to the project, and at minimum the two latest minor versions; and (2) all minor versions of NumPy released in the 24 months prior to the project, and at minimum the last three minor versions.  In ``setup.py``, the ``python_requires`` variable should be set to the minimum supported version of Python.  All supported minor versions of Python should be in the test matrix and have binary artifacts built for the release.  Minimum Python and NumPy version support should be adjusted upward only on major and minor releases, and never on a patch release.
```
Comment (jtniehof) [inline standards.md]: Can't this just be "projects should comply with NEP 29"?

**Comment by Cadair** [comment]
On line wrapping, I generally find the best thing is one line per sentence. It (ab)uses the git diff algorithm to highlight changes on the sentence by sentence basis.

**Comment by blalterman** [comment]
This is what I do and it even works with diffing LaTeX documents.

**Comment by asreimer** [comment]
The HDEE Step-2 are due 1 July. Is it expected that the PyHC standards will be updated before the due date?

File: standards.md [hunk 7]
```diff
@@ -4,13 +4,19 @@
...
+
+Amended during the Python in Heliophysics Community Meeting of November 2019.
 
 ---
 
-Agreed on 10-Dec-2018 by (in alpahbetical order)
+Agreed on 10-Dec-2018 by (in alphabetical order)
+
+**A. Annex** (JHU), **B. L. Alterman** (Univ. of Michigan), **A. Azari** (Univ. of Michigan), **W. Barnes** (Rice Univ.), **M. Bobra** (Stanford), **B. Cecconi** (Observartoire de Paris), **S. Christe** (NASA GSFC), **J. Coxon** (Univ. of Southampton), **A. DeWolfe** (LASP), **A. Halford** (NASA GSFC), **B. Harter** (LASP), **J. Ireland** (NASA GSFC), **J. Jahn** (SwRI), **J. Klenzing** (NASA GSFC), **M. Liu** (SunPy), **J. Mason** (NASA GSFC), **R. McGranaghan** (NASA JPL), **N. Murphy** (CfA), **S. Murray** (Trinity College Dublin), **J. Niehof** (Univ. of New Hampshire), **M.D. Nguyen** (Lomonosov Moscow State Univ.), **R. Panneton** (CU/LASP), **A. Pembroke** (NASA GSFC), **D. Pérez-Suárez** (University College London), **C. Piker** (Univ. of Iowa), **A. Roberts** (NASA GSFC), **D. Ryan** (NASA GSFC), **S. Savage** (NASA GSFC), **J. Smith** (NASA GSFC, Catholic Univ.), **D. Stansby** (Imperial College London), **J. Vandegriff** (JHU/APL), **R. S. Weigel** (George Mason University)
```
Comment (wtbarnes) [inline standards.md:15]:
```suggestion
**A. Annex** (JHU), **B. L. Alterman** (Univ. of Michigan), **A. Azari** (Univ. of Michigan), **W. Barnes** (NRC Postdoc residing at NRL), **M. Bobra** (Stanford), **B. Cecconi** (Observartoire de Paris), **S. Christe** (NASA GSFC), **J. Coxon** (Univ. of Southampton), **A. DeWolfe** (LASP), **A. Halford** (NASA GSFC), **B. Harter** (LASP), **J. Ireland** (NASA GSFC), **J. Jahn** (SwRI), **J. Klenzing** (NASA GSFC), **M. Liu** (SunPy), **J. Mason** (NASA GSFC), **R. McGranaghan** (NASA JPL), **N. Murphy** (CfA), **S. Murray** (Trinity College Dublin), **J. Niehof** (Univ. of New Hampshire), **M.D. Nguyen** (Lomonosov Moscow State Univ.), **R. Panneton** (CU/LASP), **A. Pembroke** (NASA GSFC), **D. Pérez-Suárez** (University College London), **C. Piker** (Univ. of Iowa), **A. Roberts** (NASA GSFC), **D. Ryan** (NASA GSFC), **S. Savage** (NASA GSFC), **J. Smith** (NASA GSFC, Catholic Univ.), **D. Stansby** (Imperial College London), **J. Vandegriff** (JHU/APL), **R. S. Weigel** (George Mason University)
```
